man/notifyd.8
.\" -*- nroff -*-
.TH NOTIFYD 8 "Project Cyrus" CMU
.\" 
.\" Copyright (c) 1998-2000 Carnegie Mellon University.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer. 
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. The name "Carnegie Mellon University" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For permission or any other legal
.\"    details, please contact  
.\"      Office of Technology Transfer
.\"      Carnegie Mellon University
.\"      5000 Forbes Avenue
.\"      Pittsburgh, PA  15213-3890
.\"      (412) 268-4387, fax: (412) 268-7395
.\"      tech-transfer@andrew.cmu.edu
.\"
.\" 4. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by Computing Services
.\"     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
.\"
.\" CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\" 
.\" $Id: notifyd.8,v 1.7 2003/08/09 23:43:14 rjs3 Exp $
.SH NAME
notifyd \- Cyrus notification server
.SH SYNOPSIS
.B notifyd
[
.B \-C
.I config-file
]
[
.B \-U
.I uses
]
[
.B \-T
.I timeout
]
.br
        [
.B \-D
]
.SH DESCRIPTION
.B Notifyd
is a daemon started from cyrmaster(8) that handles notification requests on
behalf of lmtpd(8).
.B Notifyd
accepts the request and performs the notification using the method
specified in the request.
.PP
Note that for
.B notifyd
to run properly you must set proto=udp in its cyrus.conf
services entry.  prefork=1 is also recommended.
.PP
.B Notifyd
reads its configuration options out of the
.IR imapd.conf (5)
file unless specified otherwise by \fB-C\fR.
The
.I notifysocket
option is used to specify the Unix domain socket to listen on for
notifications.
.SH OPTIONS
.TP
.BI \-C " config-file"
Read configuration options from \fIconfig-file\fR.
.TP
.BI \-U " uses"
The maximum number of times that the process should be used for new
connections before shutting down.  The default is 250.
.TP
.BI \-T " timeout"
The number of seconds that the process will wait for a new connection
before shutting down.  Note that a value of 0 (zero) will disable the
timeout.  The default is 60.
.TP
.BI \-D
Run external debugger specified in debug_command.
.SH NOTIFICATION METHODS
.TP
.B null
Ignore the notification request.
.TP
.B log
Log the notification to syslog (for testing).
.TP
.B mailto
Email the notification.  This method can ONLY be used in a
Sieve 'notify' action as it requires a \fImailto:\fR URL to be
specified as an \fI:option\fR.
.TP
.B zephyr
Send the notification as a zephyrgram.  If used in a Sieve 'notify'
action, additional recipients can be specified as \fI:options\fR.
.SH FILES
.TP
.B /etc/imapd.conf
Default configuration file.
.TP
.B /etc/cyrus.conf
Cyrus Master process configuration file.
.SH SEE ALSO
.PP
\fBimapd.conf(5)\fR, \fBcyrus.conf\fR, \fBcyrmaster(8)\fR, \fBlmtpd(8)\fR


debian/changelog
kolab-cyrus-imapd (2.2.13p1-0.4) unstable; urgency=low

  * Non-maintainer upload.
  * Fix pending l10n issues. Debconf translations:
    - Danish (Joe Hansen).  Closes: #633377
    - Dutch; (Jeroen Schot).  Closes: #656627
    - Polish (Michał Kułach).  Closes: #659569

 -- Christian Perrier <bubulle@debian.org>  Mon, 20 Feb 2012 07:15:16 +0100

kolab-cyrus-imapd (2.2.13p1-0.3) unstable; urgency=low

  * Remove db versions from debian/patches/99-update-autoconf.dpatch
    as well to fix wrong linking to db-4.7 on mipsel

 -- Ondřej Surý <ondrej@debian.org>  Tue, 28 Jun 2011 12:03:46 +0200

kolab-cyrus-imapd (2.2.13p1-0.2) unstable; urgency=low

  * Always set RET value (thanks to Steven Kurylo for catching that)

 -- Ondřej Surý <ondrej@debian.org>  Thu, 09 Jun 2011 10:07:59 +0200

kolab-cyrus-imapd (2.2.13p1-0.1) unstable; urgency=low

  * Non-maintainer upload.
    + Security upload
    + Per maintainers request
  * Imported Upstream version 2.2.13p1
  * Fix CVE-2011-1926: STARTTLS plaintext command injection
    vulnerability (VU#555316) (Closes: #629350)
  * Update autotools script before configuring by running autoreconf
  * Add support for linking with Berkeley DB5.1 (Closes: #621355)
  * Build against generic libdb-dev
  * Add automatic database upgrade script (pulled from cyrus-imapd-2.4)
    + Don't fail the upgrade if dbX.Y-util is not found
    + Try upgrade script in init.d script without upgraded databases
  * Update patches for 2.2.13p1 release
  * Filemode changes
  * Merge patches from cyrus-imapd-2.2
    + Replace sv_{yes,no,undef} with PL_sv_{yes,no,undef}
      (Closes: #628632)
    + Fix infinite loop in case of corrupted index files
    + Some GNU Hurd build fixes
    + Fix Perl IMAP.xs croak FTBFS
    + Fix cyrmaster -j (janitor) segfault
  * Build with dynamic version of com_err
  * Cherry-pick b6fd04c4 from upstream git to fix: Corrupt quota files
    cause data loss (Closes: #411240)
  * Recommend all dbX.Y-util

 -- Ondřej Surý <ondrej@debian.org>  Mon, 06 Jun 2011 21:59:10 +0200

kolab-cyrus-imapd (2.2.13-9) unstable; urgency=medium

  * Synced against cyrus-imapd package
    - Upgrade to use BerkeleyDB 4.7. (Closes: #524427)
    - Use quilt (Closes: #560628)
  * Enable imaps by default
  * Upload to unstable with urgency medium for the libkrb transition

 -- Mathieu Parent <sathieu@debian.org>  Thu, 28 Jan 2010 15:40:50 +0100

cyrus-imapd-2.2 (2.2.13-19) unstable; urgency=low

  * Switch to BerkeleyDB 4.7, the version OpenLDAP uses.
  * Add patch by Cristian Rigamonti to fix logcheck rules for "defaultbc
    doesn't exist" error message. (Closes: #511030)
  * Update Vietnamese debconf translation, thanks Clytie Siddall.
    (Closes: #548052)
  * Upload to unstable for the libkrb transition.

 -- Christoph Berg <myon@debian.org>  Sat, 23 Jan 2010 23:35:21 +0100

cyrus-imapd-2.2 (2.2.13-18) experimental; urgency=low

  [ Henrique de Moraes Holschuh ]
  * sieve/bc_eval.c (0025-upstream-fix-cve-2009-3235.dpatch):
    update for completeness to match the patch used by the security-team:
    use snprintf for scount, to future-proof against "int" larger than
    64 bits.

  [ Christoph Berg ]
  * Add myself to Uploaders.
  * Upgrade to use BerkeleyDB 4.8. (Closes: #421942)
  * Convert to use quilt, and update the patch headers to use clean paths.
    (Closes: #563303)
  * Add patch by Mathieu Parent to fix conflicting getline definition.
    (Closes: #552865)

 -- Christoph Berg <myon@debian.org>  Fri, 22 Jan 2010 22:16:02 +0100

kolab-cyrus-imapd (2.2.13-8) unstable; urgency=low

  * Remove kolab-cyrus-common.linda 
  * 01-fix_Makefile.in.dpatch: sync with cyrus-imapd-2.2
    (add '@WARNERROR@ -fPIC' to CFLAGS)
  * Disable 101-fix_Makefile.in.dpatch (redundant with above patch)
  * README.source: describe "cyrus-to-kolab" process
  * 26-local-getline.dpatch: Fix "FTBFS: ../makedepend/def.h:164: error:
     conflicting types for 'getline'" (Closes: #552848)

 -- Mathieu Parent <sathieu@debian.org>  Wed, 11 Nov 2009 23:26:07 +0100

kolab-cyrus-imapd (2.2.13-7) unstable; urgency=high

  * Setting urgency to high due to the CVE below
  * cyrus.conf: Fix "sieve port is not available form other hosts"
    (Closes: #549597)

 -- Mathieu Parent <sathieu@debian.org>  Sun, 04 Oct 2009 19:41:13 +0200

kolab-cyrus-imapd (2.2.13-6) unstable; urgency=low

  * Synced against cyrus-imapd package
  * Fix and acknowledge NMU for "CVE-2009-2632: Buffer overflow in the SIEVE
    script component" (Closes: #547712)
  * Added me as uploader
  * debian/control:
    - Prefix packages descriptions by "Kolab " 
    - Conflicts, Replaces and Provides corresponding cyrus-imapd-2.2 packages
    - Added Recommends and Suggests (based on cyrus-imapd-2.2) 
  * Get closer to cyrus-imapd-2.2: cyrus.conf 

 -- Mathieu Parent <sathieu@debian.org>  Sun, 04 Oct 2009 14:56:07 +0200

cyrus-imapd-2.2 (2.2.13-17) unstable; urgency=high

  * Security Update: CVE-2009-3235:
    Multiple stack-based buffer overflows in the Sieve parsing code,
    patches taken from upstream CVS (closes: #547947)

 -- Henrique de Moraes Holschuh <hmh@debian.org>  Tue, 22 Sep 2009 17:17:17 -0300

cyrus-imapd-2.2 (2.2.13-16) unstable; urgency=high

  * Justification for high urgency: Get security update from -15 into
    testing as fast as possible since -15 was removed before it could
    migrate.
  * Fix build dependency on tcl-dev (keeping old tcl8.3-dev to make
    backports easier) (Closes: 545281) 
  * Fix up README.Debian.database (Closes: 543293)
  * Build against newer cyrus-sasl (Closes: 527495,528468)
  * Fix cyrus-common postinst to only run makedirs if a new upstream
    is installed (Closes: 545512), add a README to note that the
    postinst needs to be changed with each new upstream
  * Fix find syntax in cyrus-makedirs (Closes: 545554)

 -- Sven Mueller <sven@debian.org>  Tue, 08 Sep 2009 10:04:14 +0200

cyrus-imapd-2.2 (2.2.13-15) unstable; urgency=high

  [Security Update]

notifyd/notifyd.c
/* notifyd.c -- main file for notifyd (notify script notification program)
 * Ken Murchison
 */
/*
 * Copyright (c) 1998-2003 Carnegie Mellon University.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The name "Carnegie Mellon University" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For permission or any other legal
 *    details, please contact  
 *      Office of Technology Transfer
 *      Carnegie Mellon University
 *      5000 Forbes Avenue
 *      Pittsburgh, PA  15213-3890
 *      (412) 268-4387, fax: (412) 268-7395
 *      tech-transfer@andrew.cmu.edu
 *
 * 4. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by Computing Services
 *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
 *
 * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
 * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
 * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
 * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * $Id: notifyd.c,v 1.19 2005/04/13 15:42:05 shadow Exp $
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/un.h>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#include <signal.h>
#include <string.h>

#include "notifyd.h"

#include "exitcodes.h"
#include "global.h"
#include "libconfig.h"
#include "xmalloc.h"


/* global state */
const int config_need_data = 0;

static int soc = 0; /* master has handed us the port as stdin */

static notifymethod_t *default_method;	/* default method daemon is using */


/* Cleanly shut down and exit */
void shut_down(int code) __attribute__ ((noreturn));
void shut_down(int code)
{
    cyrus_done();

    /* done */
    exit(code);
}

char *fetch_arg(char *head, char* tail)
{
    char *cp;

    for (cp = head; *cp && cp < tail; cp++);
    return (cp == tail ? NULL : cp + 1);
}

#define NOTIFY_MAXSIZE 8192

int do_notify()
{
    struct sockaddr_un sun_data;
    socklen_t sunlen = sizeof(sun_data);
    char buf[NOTIFY_MAXSIZE+1], *cp, *tail;
    int r, i;
    char *method, *class, *priority, *user, *mailbox, *message;
    char **options = NULL;
    long nopt;
    char *reply;
    notifymethod_t *nmethod;

    while (1) {
	method = class = priority = user = mailbox = message = reply = NULL;
	nopt = 0;

	if (signals_poll() == SIGHUP) {
	    /* caught a SIGHUP, return */
	    return 0;
	}
	r = recvfrom(soc, buf, NOTIFY_MAXSIZE, 0,
		     (struct sockaddr *) &sun_data, &sunlen);
	if (r == -1) {
	    return (errno);
	}
	buf[r] = '\0';

	tail = buf + r - 1;

	/*
	 * parse request of the form:
	 *
	 * method NUL class NUL priority NUL user NUL mailbox NUL
	 *   nopt NUL N(option NUL) message NUL
	 */
	method = (cp = buf);

	if (cp) class = (cp = fetch_arg(cp, tail));
	if (cp) priority = (cp = fetch_arg(cp, tail));
	if (cp) user = (cp = fetch_arg(cp, tail));
	if (cp) mailbox = (cp = fetch_arg(cp, tail));

	if (cp) cp = fetch_arg(cp, tail); /* skip to nopt */
	if (cp) nopt = strtol(cp, NULL, 10);
	if (nopt < 0 || errno == ERANGE) cp = NULL;

	if (cp && (nopt > 0) &&
	    !(options = (char**) xrealloc(options, nopt * sizeof(char*)))) {
	    fatal("xmalloc(): can't allocate options", EC_OSERR);
	}

	for (i = 0; cp && i < nopt; i++) {
	    options[i] = (cp = fetch_arg(cp, tail));
	}

	if (cp) message = (cp = fetch_arg(cp, tail));

	if (!message) {
	    syslog(LOG_ERR, "malformed notify request");
	    return 0;
	}

	if (!*method)
	    nmethod = default_method;
	else {
	    nmethod = methods;
	    while (nmethod->name) {
		if (!strcasecmp(nmethod->name, method)) break;
		nmethod++;
	    }
	}

	syslog(LOG_DEBUG, "do_notify using method '%s'",
	       nmethod->name ? nmethod->name: "unknown");

	if (nmethod->name) {
	    reply = nmethod->notify(class, priority, user, mailbox,
				    nopt, options, message);
	}
#if 0  /* we don't care about responses right now */
	else {
	    reply = strdup("NO unknown notification method");
	    if (!reply) {
		fatal("strdup failed", EC_OSERR);
	    }
	}
#endif

	if (reply) free(reply);
    }

    /* never reached */
    return 0;
}


void fatal(const char *s, int code)
{
    static int recurse_code = 0;

    if (recurse_code) {
	/* We were called recursively. Just give up */
	exit(recurse_code);
    }
    recurse_code = code;

    syslog(LOG_ERR, "Fatal error: %s", s);

    shut_down(code);
}

void printstring(const char *s __attribute__((unused)))
{
    /* needed to link against annotate.o */
    fatal("printstring() executed, but its not used for notifyd!",
	  EC_SOFTWARE);
}

void usage(void)
{
    syslog(LOG_ERR, "usage: notifyd [-C <alt_config>]");
    exit(EC_USAGE);
}

int service_init(int argc, char **argv, char **envp __attribute__((unused)))
{
    int opt;
    char *method = "null";

    if (geteuid() == 0) fatal("must run as the Cyrus user", EC_USAGE);

    while ((opt = getopt(argc, argv, "m:")) != EOF) {
	switch(opt) {
	case 'm':
	    method = optarg;
	    break;
	default:
	    usage();
	}
    }

    default_method = methods;
    while (default_method->name) {
	if (!strcasecmp(default_method->name, method)) break;
	default_method++;
    }

    if (!default_method) fatal("unknown notification method %s", EC_USAGE);

    signals_set_shutdown(&shut_down);

    return 0;
}

/* Called by service API to shut down the service */
void service_abort(int error)
{
    shut_down(error);
}

int service_main(int argc __attribute__((unused)),
		 char **argv __attribute__((unused)),
		 char **envp __attribute__((unused)))
{
    int r = 0;

    r = do_notify();

    shut_down(r);
    return 0;
}
